service/src/IOUring.hpp

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright (c) 2015 - 2022, Intel Corporation
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#ifndef IOURING_HPP_INCLUDE
+#define IOURING_HPP_INCLUDE
+
+#include <cstdint>
+#include <memory>
+#include <functional>
+
+namespace geopm
+{
+    class IOUring
+    {
+        public:
+            /// @brief Create and initialize an IO uring.
+            IOUring() = default;
+
+            /// @brief Clean up the IO uring.
+            virtual ~IOUring() = default;
+
+            /// @brief Submit all prepared uring operations in a batch, and
+            ///        wait for all operations to return a result. Throws if
+            ///        there are errors interacting with the completion queue.
+            ///        Failures of batched operations are reported by @p ret
+            ///        from the operation's respective prep_...  function call,
+            ///        and do not cause this function to throw.
+            virtual void submit() = 0;
```

Comment:
It might be possible for our restricted use case. Liburing does a lot of work to make sure operations are atomic with respect to the prepare/submit/get interfaces, considering challenges in out-of-order execution and context switching. Since our usage only starts filling a queue when there are no submissions in progress, and since we only need to read results after a submission fully completes, it may be possible to manually interact with the queues and with io_uring_enter in a way that lets us re-use a pre-crafted submission queue.

I think the risk of making mistakes (vs what's been designed/solved in liburing) and the relatively low payoff opportunity (can't help with batch writes which we often schedule on demand, and it looks like queue-crafting is only 10-30% of time in some few-hundred-signal reads I tried out), I don't think we should try that unless we have a specific performance objective we think it could help us meet.